File: solutions/TypeScript/[1790_Easy] Check if One String Swap Can Make Strings Equal/areAlmostEqual.ts
function areAlmostEqual(s1: string, s2: string): boolean {
  const diff: number[] = [];

  for (let i = 0; i < s1.length; i++) {
    if (s1[i] !== s2[i]) {
      diff.push(i);
    }

    if (diff.length > 2) {
      return false;
    }
  }

  if (diff.length < 2) {
    return !Boolean(diff.length);
  }

  return s1[diff[0]] === s2[diff[1]] && s1[diff[1]] === s2[diff[0]];
}

File: solutions/Java/[1790_Easy] Check if One String Swap Can Make Strings Equal/areAlmostEqual.java
class Solution {
  public boolean areAlmostEqual(String s1, String s2) {
    ArrayList<Integer> diff = new ArrayList<>();

    for (int i = 0; i < s1.length(); i++) {
      if (s1.charAt(i) != s2.charAt(i)) {
        diff.add(i);
      }

      if (diff.size() > 2) {
        return false;
      }
    }

    if (diff.size() == 0) {
      return true;
    }

    if (diff.size() == 1) {
      return false;
    }

    return (
      s1.charAt(diff.get(0)) == s2.charAt(diff.get(1))
      && s1.charAt(diff.get(1)) == s2.charAt(diff.get(0))
    );
  }
}
